pagefaultgames/pokerogue
# [Ability] Implement Illusion

## What are the changes?

Adding illusion ability, works with fusion also

## Why am I doing these changes?



It's missing

## What did change?


-Added PreSummonAbAttr class as ability attribute.
-Illusion uses the appearance of the last conscious and available Pokémon in the party (ignoring itself), including :
Species, Form, Shininess, Pokeball, Gender
-If We encounter a wild Zorua/Zoroak he will use the illusion of a pokemon in the current biome. Depending on Zorua or Zoroark level the illusion will be evolved or not.
-If we encounter a boss Zoroark he will use the illusion of Entei, Suicune or Raiku like in the original game.
-The AI move choice will be affected by the illusion.
-The type will always be displayed as the Pokémon it is disguised as, even when affected by Soak etc.
-The illusion will disappear if the Pokemon has its ability changed or suppressed.
-Pokemon cannot transform into a Pokémon that is disguised by Illusion.
-Pokemon that are disguised by Illusion cannot Transform.
-Illusion will not activate if the last conscious Pokémon is a Terastallized Ogerpon or Terapagos.

### Screenshots/Videos


https://discord.com/channels/1125469663833370665/1176874654015684739/1257835032647831682
https://discord.com/channels/1125469663833370665/1176874654015684739/1257864381937619045
https://discord.com/channels/1125469663833370665/1176874654015684739/1257865512801538059
https://discord.com/channels/1125469663833370665/1176874654015684739/1258418548074745958

## How to test the changes?




I played Zoroark and override the enemy pokemon with Zoroark so we can see him using correct illusion
I used illusion on a fusion and with a fusion
I tried to switch with the illusion
We can also run `npm run test illusion`

Old conversations here : 
https://github.com/pagefaultgames/pokerogue/pull/2217

The link for the localization pr :
https://github.com/pagefaultgames/pokerogue-locales/pull/26

## Checklist
